src/test/java/com/microsoft/sqlserver/jdbc/NTLMConnectionTest.java

```diff
@@ -0,0 +1,689 @@
+                    // this should just generate a warning but not fail
+                    fail(e.getMessage());
+                }
+            }
+
+            // verify message was logged
+            handler.flush();
+            assertTrue(loggerContent.toString().matches(TestUtils.formatErrorMsg("R_ntlmNoTimestamp")));
+        }
+    }
+
+    /*
+     * Get Server FQDN from connection
+     */
+    private void getServerFqdn(SQLServerConnection con) throws SQLException {
+        try {
+            serverFqdn = InetAddress.getByName(con.currentConnectPlaceHolder.getServerName()).getCanonicalHostName()
+                    .toUpperCase();
+        } catch (UnknownHostException e) {
+            fail("Error getting server FQDN: " + e.getMessage());
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Get bad challenge token for testNTLMBad* tests
+     * 
+     * @param offset
+     *        if > 0 offset from beginning of token if < 0 offset from end of token
+     * @param badBytes
+     *        bad bytes to write to the challenge token
+     * @return
+     */
+    private byte[] getChallengeToken(int offset, byte[] badBytes) {
+        byte[] serverBytes = serverFqdn.getBytes(java.nio.charset.StandardCharsets.UTF_16LE);
+
+        // add 4 for MSVAVDNSCOMPUTERNAME avlen
+        int targetInfoLen = challengeTargetInfo1.length + 4 + serverBytes.length + challengeTargetInfo2.length;
+        int tokenLen = challengeToken1stPart.length + targetInfoLen;
+
+        ByteBuffer token = ByteBuffer.allocate(tokenLen).order(ByteOrder.LITTLE_ENDIAN);
+        token.put(challengeToken1stPart);
+
+        token.put(challengeTargetInfo1);
+
+        token.putShort((short) serverBytes.length);
+        token.put(serverBytes);
+
+        token.put(challengeTargetInfo2);
+
+        // update targetinfo len
+        token.position(NTLM_CHALLENGE_TARGETINFOLEN_OFFSET);
+        token.putShort((short) targetInfoLen); // len
+        token.putShort((short) targetInfoLen); // maxlen
+
+        // write bad bytes
+        if (0 <= offset) {
+            token.position(offset);
+            token.put(badBytes);
+        } else {
+            token.position(tokenLen + offset);
+            token.put(badBytes);
+        }
+
+        return token.array();
+    }
+
+    /*
+     * Test invalid properties
+     */
+    private void testInvalidProperties(String connectionString, String resourceKeyword) {
+        try (Connection con = DriverManager.getConnection(connectionString)) {
+            fail();
+        } catch (SQLException e) {
+            assertTrue(e.getMessage().matches(TestUtils.formatErrorMsg(resourceKeyword)));
+        }
+    }
+
+    /*
+     * Verify NTLM authentication
+     */
+    private void verifyNTLM(Connection con) throws SQLException {
+        try (Statement stmt = con.createStatement(); ResultSet rs = stmt
+                .executeQuery("select auth_scheme from sys.dm_exec_connections where session_id=@@spid")) {

```

Comment:
Please note that this query requires `VIEW SERVER STATE` permission. We could grant it to NTLM user using the original `connectionString` in `setUp()` **assuming** original user has `GRANT` permission.